File: Scala/src/Sources/Insertion/insertion.scala
package Sources.Insertion

import Others.swap.swap
import Others.value._

object insertion {
    private def insertion(a: Array[Int]): Unit = {
        for (i <- 1 until a.length) {
            val cm = a(i)
            var j = i - 1
            while (j >= 0 && a(j) > cm) {
                swap(a, j + 1, j)
                j -= 1
            }
        }
    }

    def main(args: Array[String]): Unit = {
        val s = 1000000 // Changeable size
        val a = als(s)

        val start = System.currentTimeMillis()

        insertion(a)

        val end = System.currentTimeMillis() - start

        for (i <- a.indices)
            println(a(i))
        print(end + " ms")
    }
}

File: Kotlin/src/main/kotlin/Sources/Insertion/insertion.kt
package Sources.Insertion

import swap

fun insertion(a: IntArray) {
    for (i in 1 until a.size) {
        val cm = a[i]
        var j = i-1
        while (j >= 0 && a[j] > cm) {
            swap(a, j+1, j)
            --j
        }
    }
}

fun main() {
    val a = intArrayOf(5, 4, 3, 2, 1)

    val start = System.currentTimeMillis()

    insertion(a)

    val end = System.currentTimeMillis() - start

    for (i in a.indices)
        println(a[i])
    print("$end ms")
}